### PR TITLE
etcd/client(ticdc): add retry operation for etcd transaction api (#4248)

### LINKED ISSUE #4248: Owner exits unexpectedly when PD leader is killed -9
### What did you do?

1. Running a changefeed.
2. Kill PD leader (also etcd leader too).

Owner node exits unexpectedly.

image

```log
[2022/01/03 22:59:44.154 +08:00] [WARN] [etcd_worker.go:405] ["Etcd transaction took too long"] [duration=11.000997285s]
[2022/01/03 22:59:44.154 +08:00] [INFO] [client.go:189] ["WatchWithChan exited"]
[2022/01/03 22:59:44.154 +08:00] [INFO] [capture.go:430] ["run owner exited"] [error="etcdserver: request timed out, possibly due to previous leader failure"] [errorVerbose="etcdserver: request timed out, possibly due to previous leader failure\ngithub.com/pingcap/errors.AddStack\n\tgithub.com/pingcap/errors@v0.11.5-0.20211009033009-93128226aaa3/errors.go:174\ngithub.com/pingcap/errors.Trace\n\tgithub.com/pingcap/errors@v0.11.5-0.20211009033009-93128226aaa3/juju_adaptor.go:15\ngithub.com/pingcap/tiflow/pkg/orchestrator.(*EtcdWorker).commitChangedState\n\tgithub.com/pingcap/tiflow/pkg/orchestrator/etcd_worker.go:409\ngithub.com/pingcap/tiflow/pkg/orchestrator.(*EtcdWorker).applyPatchGroups\n\tgithub.com/pingcap/tiflow/pkg/orchestrator/etcd_worker.go:337\ngithub.com/pingcap/tiflow/pkg/orchestrator.(*EtcdWorker).Run\n\tgithub.com/pingcap/tiflow/pkg/orchestrator/etcd_worker.go:207\ngithub.com/pingcap/tiflow/cdc/capture.(*Capture).runEtcdWorker\n\tgithub.com/pingcap/tiflow/cdc/capture/capture.go:450\ngithub.com/pingcap/tiflow/cdc/capture.(*Capture).campaignOwner\n\tgithub.com/pingcap/tiflow/cdc/capture/capture.go:428\ngithub.com/pingcap/tiflow/cdc/capture.(*Capture).run.func2\n\tgithub.com/pingcap/tiflow/cdc/capture/capture.go:292\nruntime.goexit\n\truntime/asm_amd64.s:1581"]
[2022/01/03 22:59:44.156 +08:00] [INFO] [capture.go:293] ["the owner routine has exited"] [error="etcdserver: request timed out, possibly due to previous leader failure"] [errorVerbose="etcdserver: request timed out, possibly due to previous leader failure\ngithub.com/pingcap/errors.AddStack\n\tgithub.com/pingcap/errors@v0.11.5-0.20211009033009-93128226aaa3/errors.go:174\ngithub.com/pingcap/errors.Trace\n\tgithub.com/pingcap/errors@v0.11.5-0.20211009033009-93128226aaa3/juju_adaptor.go:15\ngithub.com/pingcap/tiflow/pkg/orchestrator.(*EtcdWorker).commitChangedState\n\tgithub.com/pingcap/tiflow/pkg/orchestrator/etcd_worker.go:409\ngithub.com/pingcap/tiflow/pkg/orchestrator.(*EtcdWorker).applyPatchGroups\n\tgithub.com/pingcap/tiflow/pkg/orchestrator/etcd_worker.go:337\ngithub.com/pingcap/tiflow/pkg/orchestrator.(*EtcdWorker).Run\n\tgithub.com/pingcap/tiflow/pkg/orchestrator/etcd_worker.go:207\ngithub.com/pingcap/tiflow/cdc/capture.(*Capture).runEtcdWorker\n\tgithub.com/pingcap/tiflow/cdc/capture/capture.go:450\ngithub.com/pingcap/tiflow/cdc/capture.(*Capture).campaignOwner\n\tgithub.com/pingcap/tiflow/cdc/capture/capture.go:428\ngithub.com/pingcap/tiflow/cdc/capture.(*Capture).run.func2\n\tgithub.com/pingcap/tiflow/cdc/capture/capture.go:292\nruntime.goexit\n\truntime/asm_amd64.s:1581"]
…
[2022/01/03 22:59:59.836 +08:00] [INFO] [capture.go:317] ["the processor routine has exited"] [error="[CDC:ErrPDEtcdAPIError]etcd api call error: context canceled"] [errorVerbose="[CDC:ErrPDEtcdAPIError]etcd api call error: context canceled\ngithub.com/pingcap/errors.AddStack\n\tgithub.com/pingcap/errors@v0.11.5-0.20211009033009-93128226aaa3/errors.go:174\ngithub.com/pingcap/errors.(*Error).GenWithStackByArgs\n\tgithub.com/pingcap/errors@v0.11.5-0.20211009033009-93128226aaa3/normalize.go:159\ngithub.com/pingcap/tiflow/pkg/errors.WrapError\n\tgithub.com/pingcap/tiflow/pkg/errors/helper.go:30\ngithub.com/pingcap/tiflow/cdc/capture.(*Capture).runEtcdWorker\n\tgithub.com/pingcap/tiflow/cdc/capture/capture.go:464\ngithub.com/pingcap/tiflow/cdc/capture.(*Capture).run.func3\n\tgithub.com/pingcap/tiflow/cdc/capture/capture.go:316\nruntime.goexit\n\truntime/asm_amd64.s:1581"]
```

### What did you expect to see?

A TiCDC node should never exit unexpectedly.

### What did you see instead?

A TiCDC node exits unexpectedly.

### Versions of the cluster

Upstream TiDB cluster version (execute `SELECT tidb_version();` in a MySQL client):

```console
release-5.4
```

TiCDC version (execute `cdc version`):

```console
e274ab691d3b1bbd21e4d0e52e512003762b1767
```

### PR DESCRIPTION
**Pull request by CharlesCheung96**
### What problem does this PR solve?

Issue Number: close #4248 

### What is changed and how it works?
1. Add retry strategy for etcd transaction, which uniformly handles errors related to etcd transactions and only throws errors that are not retryable, such as context.Canceled, context.DeadlineExceeded, errors.ErrReachMaxTry.
2. Catch all of the EtcdError in the etcd_worker and retry the Txn operation again to prevent the cdc process from exiting abnormally. 
3. When the connection between cdc and etcd is abnormal, session will be responsible for disconnecting and throwing ErrEtcdSessionDone error, which causes the capture to [automatically recover after suicide.](https://github.com/pingcap/tiflow/blob/master/cdc/capture/capture.go#L467)

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change
 - Has exported variable/fields change

Side effects

 - Possible performance regression
 - Increased code complexity

Related changes

 - Need to cherry-pick to the release branch

### Release note 

```release-note
Fix a bug that owner exits abnormally when PD leader is killed
```

### CODE REVIEW

File: pkg/etcd/client.go [hunk 1]
```diff
@@ -145,12 +159,31 @@ func isRetryableError(rpcName string) retry.IsRetryable {
...
 		if rpcName == EtcdRevoke {
-			if etcdErr, ok := err.(rpctypes.EtcdError); ok && etcdErr.Code() == codes.NotFound {
+			if etcdErr, ok := err.(v3rpc.EtcdError); ok && etcdErr.Code() == codes.NotFound {
 				// it means the etcd lease is already expired or revoked
 				return false
 			}
 		}
 
+		if rpcName == EtcdTxn {
+			switch err {
```
Comment (CharlesCheung96) [inline pkg/etcd/client.go:169]: ref https://github.com/pingcap/tiflow/blob/master/dm/pkg/etcdutil/etcdutil.go#L51-L66

File: pkg/etcd/client.go [hunk 2]
```diff
@@ -145,12 +159,31 @@ func isRetryableError(rpcName string) retry.IsRetryable {
 			return false
 		}
 		if rpcName == EtcdRevoke {
-			if etcdErr, ok := err.(rpctypes.EtcdError); ok && etcdErr.Code() == codes.NotFound {
+			if etcdErr, ok := err.(v3rpc.EtcdError); ok && etcdErr.Code() == codes.NotFound {
 				// it means the etcd lease is already expired or revoked
 				return false
 			}
 		}
```
Comment (CharlesCheung96) [inline pkg/etcd/client.go:167]: ref https://github.com/pingcap/tiflow/blob/master/dm/pkg/etcdutil/etcdutil.go#L51-L66

File: pkg/orchestrator/etcd_worker.go [hunk 3]
```diff
@@ -257,6 +254,13 @@ func (worker *EtcdWorker) Run(ctx context.Context, session *concurrency.Session,
 	}
 }
 
+func isRetryableError(err error) bool {
+	err = errors.Cause(err)
+	return cerrors.ErrEtcdTryAgain.Equal(err) ||
+		cerrors.ErrReachMaxTry.Equal(err) ||
+		context.DeadlineExceeded == err
```
Comment (CharlesCheung96) [inline pkg/orchestrator/etcd_worker.go:261]: I'm not sure if the `cerrors.ErrReachMaxTry and context.DeadlineExceeded `should handle here.

**Comment by liuzix** [comment]
Please add unit tests

File: pkg/etcd/client.go [hunk 4]
```diff
@@ -51,6 +51,15 @@ const (
...
+	// etcdTxnTimeoutDuration represents the timeout duration for committing a
+	// transaction to Etcd
+	etcdTxnTimeoutDuration = 30 * time.Second
+)
+
+var (
+	TxnEmptyCmps    = []clientv3.Cmp{}
+	TxnEmptyOpsThen = []clientv3.Op{}
+	TxnEmptyOpsElse = []clientv3.Op{}
 )
```
Comment (amyangfei) [inline pkg/etcd/client.go:63]: We can move these variables to `client_test.go` since they are used in test only.

File: pkg/etcd/client.go (hunk 2 above)
Comment (amyangfei) [reply pkg/etcd/client.go:167]: We can extract a function, such as `EtcdRetryableError() bool` in the future, similar to the work in https://github.com/pingcap/tiflow/issues/3242

File: pkg/etcd/client.go (hunk 4 above)
Comment (CharlesCheung96) [reply pkg/etcd/client.go:63]: We also use it in etcd_worker and etcd.go, such as
https://github.com/pingcap/tiflow/pull/4474/files#diff-1f744cd3130fdaa4293a04fa689c341aa1086bc64278c7a5af46e7774dc16cd9R389

**Comment by amyangfei** [comment]
Has the case in #4248 been tested manually
- How long will PD recover after kill -9
- What is the behavior of TiCDC after PD is killed

File: pkg/etcd/client.go (hunk 2 above)
Comment (CharlesCheung96) [reply pkg/etcd/client.go]: Done, PTAL

**Review by amyangfei — APPROVED** [review]

File: pkg/etcd/client.go [hunk 5]
```diff
@@ -145,12 +160,14 @@ func isRetryableError(rpcName string) retry.IsRetryable {
 			return false
 		}
 		if rpcName == EtcdRevoke {
```
Comment (asddongmen) [inline pkg/etcd/client.go:162]: May be `swtich case` here is better.

Comment (asddongmen) [reply pkg/etcd/client.go:162]: > May be `swtich case` here is better.
rest LGTM

Comment (CharlesCheung96) [reply pkg/etcd/client.go]: Done, PTAL

**Review by asddongmen — APPROVED** [review]

**Comment by asddongmen** [comment]
Please refine release note.

**Pull request merged**